src/mercury_engine_data_structures/formats/rom3ds.py

```diff
@@ -0,0 +1,342 @@
+    # there is a lot of card data, test data etc. totally irrelevant see https://www.3dbrew.org/wiki/NCSD
+    "_garbage" / Seek(construct.this.ncsd.partitions[0].offset * 0x200),
+    "rom_struct" / ROMStructure,
+)
+
+AddCia = Struct(
+    "header_size" / Int32ul,
+    "type" / Int16ul,
+    "format_version" / Int16ul,
+    "certificate_size" / Int32ul,
+    "ticket_size" / Int32ul,
+    "tmd_size" / Int32ul,
+    "footer_size" / Int32ul,
+    "content_size" / Int32ul,
+    # FIXME: there must be an easier way to skip over these aligned sections
+    "_return_to_start" / Seek(0),
+    "_garbage"
+    / AlignedStruct(
+        64,
+        "header" / Bytes(construct.this._.header_size),
+        "certificate" / Bytes(construct.this._.certificate_size),
+        "ticket" / Bytes(construct.this._.ticket_size),
+        "tmd" / Bytes(construct.this._.tmd_size),
+    ),
+    "rom_struct" / ROMStructure,
+)
+
+AddCxi = Struct(
+    "rom_struct" / ROMStructure,
+)
+
+LzssFooter = Struct(
+    "off_size_comp" / Int32ul,  # 0xOOSSSSSS, where O == reverse offset and S == size
+    "addsize_dec" / Int32ul,  # decompressed size - compressed size
+)
+
+RomFsDirectoryEntry = Struct(
+    "parent_offset" / Int32ul,
+    "sibling_offset" / Int32ul,
+    "child_offset" / Int32ul,
+    "file_offset" / Int32ul,
+    "hash_sibling_offset" / Int32ul,
+    "name" / PascalString(Int32ul, "utf_16_le"),
+)
+
+RomFsFileEntry = Struct(
+    "parent_offset" / Int32ul,
+    "sibling_offset" / Int32ul,
+    "data_offset" / Int64ul,
+    "data_size" / Int64ul,
+    "hash_sibling_offset" / Int32ul,
+    "name" / PascalString(Int32ul, "utf_16_le"),
+)
+
+
+class DirectoryEntry:
+    def __init__(self, parent_path: str, entry: construct.Container):
+        self._entry = entry
+        if parent_path == "":
+            self.complete_path = self._entry.name
+        else:
+            self.complete_path = parent_path + "/" + self._entry.name
+
+
+class FileEntry:
+    def __init__(self, parent_path: str, entry: construct.Container):
+        self._entry = entry
+        if parent_path == "":
+            self.complete_path = self._entry.name
+        else:
+            self.complete_path = parent_path + "/" + self._entry.name
+        self.data_size = entry.data_size
+        self.data_offset = entry.data_offset
+
+
+class Rom3DS:
+    def __init__(self, file_name: str, file_stream: BufferedReader):
+        self.file_name_to_entry = {}
+        self.file_stream = file_stream
+        try:
+            if file_name.lower().endswith("cia"):
+                self.raw = AddCia.parse_stream(file_stream)
+            elif file_name.lower().endswith("3ds"):
+                self.raw = Add3ds.parse_stream(file_stream)
+            elif file_name.lower().endswith("app") or file_name.lower().endswith("cxi"):
+                self.raw = AddCxi.parse_stream(file_stream)
+            else:
+                raise ValueError('Input does not end with ".cia", ".3ds", ".app" or ".cxi"')
```

Comment:
I'd change this class to just receive the `raw` and `file_stream` directly, and have a helper function that knows how to identify which kind of construct they need.